File: java/src/org/ggolawski/security/utils/crypto/DigitalSigning.java
package org.ggolawski.security.utils.crypto;

import java.security.GeneralSecurityException;
import java.security.KeyPair;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;

import org.apache.commons.codec.binary.Hex;

public class DigitalSigning {
	public static void main(String[] args) throws GeneralSecurityException {
		String text = "Very secret information";
		System.out.println("Beginning; message to sign: " + text);
		
		// Generate public and private keys
		System.out.println("Generating key pair...");
	    KeyPair keyPair = Utils.generateKeyPair("EC", 256);
	    PublicKey publicKey = keyPair.getPublic();
	    PrivateKey privateKey = keyPair.getPrivate();
	    System.out.println("Public key: " + Hex.encodeHexString(publicKey.getEncoded()));
	    System.out.println("Private key: " + Hex.encodeHexString(privateKey.getEncoded()));
	    
	    Signature signature = Signature.getInstance("SHA256withECDSA");
	    
	    // Sign the message
	    System.out.println("Signing...");
	    signature.initSign(privateKey);
	    signature.update(text.getBytes());
	    byte[] messageSignature = signature.sign();
	    System.out.println("Signature: " + Hex.encodeHexString(messageSignature));
	    
	    // Alter the signature
//	    messageSignature[10] = 33;
	    
	    // Verify the signature
	    System.out.println("Verifying...");
	    signature.initVerify(publicKey);
	    signature.update(text.getBytes());
	    boolean valid = signature.verify(messageSignature);
	    System.out.println("Signature valid? " + valid);
	}
}

File: php/crypto/DigitalSigning.php
<?php
$text = "Very secret information";
echo "Beginning; message to sign: " . $text, PHP_EOL;

// Generate public and private keys
echo "Generating key pair...", PHP_EOL;
$config = array(
    "curve_name" => 'prime256v1',
    "private_key_type" => OPENSSL_KEYTYPE_EC,
);
$res = openssl_pkey_new($config);
$public_key = openssl_pkey_get_details($res)["key"];
openssl_pkey_export($res, $private_key);
echo "Public key: " . $public_key, PHP_EOL;
echo "Private key: " . $private_key, PHP_EOL;

// Sign the message
echo "Signing...", PHP_EOL;
openssl_sign($text, $signature, $private_key, OPENSSL_ALGO_SHA256);
echo "Signature: " . bin2hex($signature), PHP_EOL;

// Alter the signature
// $signature[10] = 33;

// Verify the signature
echo "Verifying...", PHP_EOL;
$valid = openssl_verify($text, $signature, $public_key, OPENSSL_ALGO_SHA256);
echo "Signature valid? " . $valid, PHP_EOL;
?>